pocket_coffea/lib/leptons.py

```diff
@@ -1,5 +1,74 @@
 import awkward as ak
 import numpy as np
+import correctionlib
+
+
+def get_ele_scaled(ele, jsonFileName, isMC, runNr):
+    evaluator = correctionlib.CorrectionSet.from_file(jsonFileName)
+    evaluator_scale = evaluator["Scale"]
+    ele_gain_flat = ak.flatten(ele["seedGain"])
+    ele_eta_flat = ak.flatten(ele["etaSC"])
+    ele_r9_flat = ak.flatten(ele["r9"])
+    ele_pt_flat = ak.flatten(ele["pt"])
+    ele_counts = ak.num(ele["pt"])
+    ele_runNr_flat = np.repeat(runNr, ele_counts)
+    if not isMC: # for data return nominal
+        scale = evaluator_scale.evaluate(
+            "total_correction", 
+            ele_gain_flat, 
+            ele_runNr_flat, 
+            ele_eta_flat, 
+            ele_r9_flat, 
+            ele_pt_flat
+        )
+        ele_pt_corrected = ak.unflatten(scale * ele_pt_flat, counts=ele_counts)
+        return {"nominal": ele_pt_corrected}
+    else: # for MC return shifts
+        scale_MC_unc = evaluator_scale.evaluate(
+            "total_uncertainty", 
+            ele_gain_flat, 
+            ele_runNr_flat, 
+            ele_eta_flat, 
+            ele_r9_flat, 
+            ele_pt_flat
+        )
+        ele_pt_scale_up = ak.unflatten((1+scale_MC_unc) * ele_pt_flat, counts=ele_counts)
+        ele_pt_scale_down = ak.unflatten((1-scale_MC_unc) * ele_pt_flat, counts=ele_counts)
+        return {"Up": ele_pt_scale_up, "Down": ele_pt_scale_down}
+
+def get_ele_smeared(mc_ele, jsonFileName, isMC, nominal=True):
+    
+    if not isMC:
+        return
+    evaluator = correctionlib.CorrectionSet.from_file(jsonFileName)
+    evaluator_smearing = evaluator["Smearing"]
+    ele_eta_flat = ak.flatten(mc_ele["etaSC"])
+    ele_r9_flat = ak.flatten(mc_ele["r9"])
+    ele_pt_flat = ak.flatten(mc_ele["pt"])
+    ele_counts = ak.num(mc_ele["pt"])
+    rng = np.random.default_rng(seed=125)
```

Comment:
I agree that this needs to be deterministic, but like this the smearning factor is the same at each function invocation. 

Maybe we can create a seed from the dataset/file name? So that it is still reproducibile but different for different files?

Reply:
Do you have a suggestion on how to do this?
Is the function called once per dataset/file, or once per chunk?

Reply:
The function is called once per chunk. so it can be called multiple time per file. 
We can make it dependent only on the filename or on the filename and chunk.

We may use the filename and entrystart to define a seed. Something like
 ```seed = hash(self.events.metadata['fileuuid'])+self.events.metadata['entrystart']```

The result would be reproducible in case the analysis is re-run with the same chunksize. But this is expected with smearing operations. I think it is already nice to be able to reproduce exactly the same numbers if one does not change the file splitting. 

What do you think?